src/Microsoft.AspNet.Server.Kestrel/Http/SocketInput.cs

```diff
@@ -194,11 +194,13 @@ public void OnCompleted(Action continuation)
             }
             else if (awaitableState == _awaitableIsCompleted)
             {
-                ThreadPool.QueueUserWorkItem((o) => ((Action)o)(), continuation);
+                // Read already completed, continue on calling thread
+                continuation();
```

Comment:
@stephentoub is there a way to detect if you are already on threadpool/top of stack/async complete; rather than doing a double queue? E.g. the stack trace on https://github.com/aspnet/KestrelHttpServer/issues/451 is clearly top of stack.